### ISSUE TITLE
S1541 and null conditional operator, c#, take 2

### ISSUE THREAD

**Issue by Andrioden** [SonarSource/sonar-dotnet#6255]
### Description

I want another take on https://github.com/SonarSource/sonar-dotnet/issues/1585 . Null conditional operator is not a branch. Even if you can argue that it is on compiled code, it is not in anyone's mental model of the code.

Your reasoning

> The null-conditional being a syntax sugar for a if not null we do increase the count (i.e. you created one more branch).

is imo wrong, this is a **helper method** for "return null if null else value" made available through syntax sugar.

If you stand your ground on your reasoning, i would kindly request that you either:
- add some sort of rule configuration that allows one to ignore `null conditional operator`
- or add a new rule like `S1541b` which dont count null conditional operator

### Repro steps

1. Create a deep class hierarchi
2. Call it with
```csharp
var classA = new ClassA();
var a = ClassA?.VarB?.VarC?.VarD?.VarE?.VarF
var b = ClassA?.VarB?.VarC?.VarD?.VarE?.VarF?.VarG
```

You have now created too complex code (11 complexity score).

### Expected behavior

No S1541 breach

### Actual behavior

Breach of S1541

### Known workarounds

One workaround, to illustrate the sillyness of this rule, is that we could for each class create our own extension helper method

```csharp
    public static class MyExtensions
    {
        public static ClassB VarBHelper(this ClassA classA)
        {
            return classA?.VarB;
        }
    }

```

Then call them like this
```csharp
var classA = new ClassA();
var a = classA.VarBHelper()
```

This would count as 0 complexity

**Comment by costin-zaharia-sonarsource** [comment]
Hi @Andrioden,

Thanks for your feedback.

I do believe that the current behavior is the correct one.

> Cyclomatic complexity is a [software metric](https://en.wikipedia.org/wiki/Software_metric) used to indicate the [complexity of a program](https://en.wikipedia.org/wiki/Programming_complexity). It is a quantitative measure of the number of linearly independent paths through a program's [source code](https://en.wikipedia.org/wiki/Source_code).

Source: https://en.wikipedia.org/wiki/Cyclomatic_complexity

Did you try to use [cognitive complexity rule](https://rules.sonarsource.com/csharp/RSPEC-3776) instead? You can find the whitepaper here: https://www.sonarsource.com/docs/CognitiveComplexity.pdf
